airbyte-integrations/connectors/source-oracle-v2/src/cdk/main/kotlin/io/airbyte/cdk/command/ConnectorCommandLinePropertySource.kt

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.command
+
+import io.airbyte.cdk.integrations.base.JavaBaseConstants
+import io.airbyte.cdk.operation.CONNECTOR_OPERATION
+import io.airbyte.cdk.operation.OperationType
+import io.github.oshai.kotlinlogging.KotlinLogging
+import io.micronaut.context.env.MapPropertySource
+import io.micronaut.core.cli.CommandLine
+import java.io.File
+import java.nio.file.Path
+
+private val log = KotlinLogging.logger {}
+
+/**
+ * Custom Micronaut [MapPropertySource] that reads the command line arguments provided via the
+ * connector CLI and turns them into configuration properties. This allows the arguments to be
+ * injected into code that depends on them via Micronaut.
+ */
+class ConnectorCommandLinePropertySource(commandLine: CommandLine) :
+    MapPropertySource("connector", resolveValues(commandLine))
+
+const val CONNECTOR_CONFIG_PREFIX: String = "airbyte.connector.config"
+const val CONNECTOR_CATALOG_PREFIX: String = "airbyte.connector.catalog"
+const val CONNECTOR_STATE_PREFIX: String = "airbyte.connector.state"
+const val CONNECTOR_OUTPUT_FILE = "airbyte.connector.output.file"
+
+private fun resolveValues(commandLine: CommandLine): Map<String, Any> {
+    val ops: List<OperationType> =
+        OperationType.entries.filter { commandLine.optionValue(it.name.lowercase()) != null }
+    if (ops.isEmpty()) {
+        throw IllegalArgumentException("Command line is missing an operation.")
+    }
+    if (ops.size > 1) {
+        throw IllegalArgumentException("Command line has multiple operations: $ops")
+    }
+    val values: MutableMap<String, Any> = mutableMapOf()
+    values[CONNECTOR_OPERATION] = ops.first().name.lowercase()
+    commandLine.optionValue("output")?.let { values[CONNECTOR_OUTPUT_FILE] = it }
+    for ((cliOptionKey, prefix) in
+        mapOf(
```

Comment:
looks like this should be inverted: Go through the CLI parameters, and throw if there's anything we don't know about, no?

Reply:
imo we should be permissive with cli params - it makes adding new params easier, since older connector versions will just ignore them

Reply:
I guess that's a good point. It raises another question to me: how do we want to pass additional options to a connector? CLI parameters, config.json, env variables, properties file ? I feel like the CLI parameters is the worst of those since everything is inside a docker container

Reply:
but yeah, if we add CLI parameters in the docker, we're kind of screwed if we're not permissive...

Reply:
Picocli does this CLI arg validation for us and throws when it violates the grammar. In fact, picocli is so good at this, it's why I used the programmatic CLI grammar spec definition rather than the annotations. I get to allow extra args in tests (and only in tests!) that way.

Reply:
If we need to add more params, we just add them to the picocli spec in `PicocliCommandLineFactory` , it's quite easy.